File: server/gameserver/src/main/java/me/eddiep/ghost/gameserver/api/network/packets/PlayerStatePacket.java
package me.eddiep.ghost.gameserver.api.network.packets;

import me.eddiep.ghost.game.match.world.timeline.PlayableSnapshot;
import me.eddiep.ghost.gameserver.api.network.TcpUdpClient;
import me.eddiep.ghost.gameserver.api.network.TcpUdpServer;
import me.eddiep.ghost.network.packet.Packet;

import java.io.IOException;

public class PlayerStatePacket extends Packet<TcpUdpServer, TcpUdpClient> {
    public PlayerStatePacket(TcpUdpClient client) {
        super(client);
    }

    @Override
    protected void onWritePacket(TcpUdpClient client, Object... args) throws IOException {
        if (args.length != 1)
            return;

        PlayableSnapshot player = (PlayableSnapshot)args[0];

        write((byte)0x12)
            .write(client.getPlayer().getID() == player.getID() ? (short)0 : player.getID())
            .write(player.getLives())
            .write(player.isDead())
            .write(player.isFrozen())
            .endTCP();
    }
}

File: client/core/src/me/eddiep/ghost/client/network/packets/PlayerStatePacket.kt
package me.eddiep.ghost.client.network.packets

import me.eddiep.ghost.client.Ghost
import me.eddiep.ghost.client.core.sprites.NetworkPlayer
import me.eddiep.ghost.client.network.Packet
import me.eddiep.ghost.client.network.PlayerClient

class PlayerStatePacket : Packet<PlayerClient>() {

    override fun handle() {
        val id = consume(2).asShort()
        val lives = consume(1).asByte()
        val isDead = consume(1).asBoolean()
        val isFrozen = consume(1).asBoolean()
        val isInvincible = consume(1).asBoolean()

        val e = client.game.findEntity(id)

        if (e is NetworkPlayer) {
            val player : NetworkPlayer = e

            player.lives = lives
            player.dead = isDead
            player.frozen = isFrozen
        }
    }
}
